pkg/operator/boundsatokensignercontroller/controller.go

```diff
@@ -0,0 +1,296 @@
+
+	TokenReadyAnnotation = "kube-apiserver.openshift.io/ready-to-use"
+	readyInterval        = 5 * time.Minute
+
+	CertConfigMapName = "bound-sa-token-signing-certs"
+)
+
+// BoundSATokenSignerController manages the keypair used to sign bound
+// tokens and the key bundle used to verify them.
+//
+// This code is derived from the controller used by the
+// kube-controller-manager-operator to manage the key material for the
+// legacy sa token signer.
+type BoundSATokenSignerController struct {
+	secretClient    corev1client.SecretsGetter
+	configMapClient corev1client.ConfigMapsGetter
+	eventRecorder   events.Recorder
+
+	cachesSynced []cache.InformerSynced
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+func NewBoundSATokenSignerController(
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+
+) *BoundSATokenSignerController {
+
+	ret := &BoundSATokenSignerController{
+		secretClient:    v1helpers.CachedSecretGetter(kubeClient.CoreV1(), kubeInformersForNamespaces),
+		configMapClient: v1helpers.CachedConfigMapGetter(kubeClient.CoreV1(), kubeInformersForNamespaces),
+		eventRecorder:   eventRecorder.WithComponentSuffix("bound-sa-token-signer-controller"),
+
+		cachesSynced: []cache.InformerSynced{
+			kubeInformersForNamespaces.InformersFor(operatorNamespace).Core().V1().Secrets().Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Secrets().Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer().HasSynced,
+		},
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "BoundSATokenSignerController"),
+	}
+
+	kubeInformersForNamespaces.InformersFor(operatorNamespace).Core().V1().Secrets().Informer().AddEventHandler(ret.eventHandler())
+	kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Secrets().Informer().AddEventHandler(ret.eventHandler())
+	kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer().AddEventHandler(ret.eventHandler())
+
+	return ret
+}
+
+func (c *BoundSATokenSignerController) sync() error {
+	signingSecret, err := c.secretClient.Secrets(operatorNamespace).Get(NextSigningKeySecretName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	needKeypair := errors.IsNotFound(err) || len(signingSecret.Data[PrivateKeyKey]) == 0 || len(signingSecret.Data[PublicKeyKey]) == 0
+	if needKeypair {
+		newSecret, err := newSigningSecret()
+		if err != nil {
+			return err
+		}
+		signingSecret, _, err = resourceapply.ApplySecret(c.secretClient, c.eventRecorder, newSecret)
+		if err != nil {
+			return err
+		}
+		// requeue for after we should have recovered
+		c.queue.AddAfter(workQueueKey, readyInterval+10*time.Second)
```

Comment:
why this line?

Reply:
And if that line, why no immediate `return` ?

Reply:
Added comment, PTAL

Reply:
Removed